tokio-net/src/driver/sharded_slab/page/mod.rs

```diff
@@ -0,0 +1,198 @@
+        // zeros (unused twos places) in the number's binary representation, and
+        // subtracting that count from the total number of bits in a word.
+        super::WIDTH - ((self.addr + INITIAL_SIZE) >> ADDR_INDEX_SHIFT).leading_zeros() as usize
+    }
+
+    pub(super) fn offset(self) -> usize {
+        self.addr
+    }
+}
+
+impl Pack for Addr {
+    const LEN: usize = super::MAX_PAGES + ADDR_INDEX_SHIFT;
+
+    type Prev = ();
+
+    fn as_usize(&self) -> usize {
+        self.addr
+    }
+
+    fn from_usize(addr: usize) -> Self {
+        debug_assert!(addr <= Self::BITS);
+        Self { addr }
+    }
+}
+
+pub(super) type Iter<'a, T> =
+    std::iter::FilterMap<std::slice::Iter<'a, Slot<T>>, fn(&'a Slot<T>) -> Option<&'a T>>;
+
+pub(super) struct Page<T> {
+    prev_sz: usize,
+    remote_head: AtomicUsize,
+    local_head: usize,
+    slab: Box<[Slot<T>]>,
+}
+
+impl<T> Page<T> {
+    pub(super) fn new(size: usize, prev_sz: usize) -> Self {
+        let mut slab = Vec::with_capacity(size);
+        slab.extend((1..size).map(Slot::new));
+        slab.push(Slot::new(Addr::NULL));
+        Self {
+            prev_sz,
+            remote_head: AtomicUsize::new(Addr::NULL),
+            local_head: 0,
+            slab: slab.into_boxed_slice(),
+        }
+    }
+
+    #[inline]
+    pub(super) fn insert(&mut self, t: &mut Option<T>) -> Option<usize> {
+        let head = self.local_head;
+        // are there any items on the local free list? (fast path)
+        let head = if head < self.slab.len() {
+            head
+        } else {
+            // if the local free list is empty, pop all the items on the remote
+            // free list onto the local free list.
+            self.remote_head.swap(Addr::NULL, Ordering::Acquire)
+        };
+
+        // if the head is still null, both the local and remote free lists are
+        // empty --- we can't fit any more items on this page.
+        if head == Addr::NULL {
+            return None;
+        }
+
+        let slot = &mut self.slab[head];
+        let gen = slot.insert(t);
+        self.local_head = slot.next();
+        Some(gen.pack(head + self.prev_sz))
+    }
+
+    #[inline]
+    pub(super) fn get(&self, addr: Addr, idx: usize) -> Option<&T> {
+        let offset = addr.offset() - self.prev_sz;
+        self.slab
+            .get(offset)?
+            .get(slot::Generation::from_packed(idx))
+    }
+
+    pub(super) fn remove_local(&mut self, addr: Addr, gen: slot::Generation) -> Option<T> {
+        let offset = addr.offset() - self.prev_sz;
+        let val = self.slab.get(offset)?.remove(gen, self.local_head);
+        self.local_head = offset;
+        val
+    }
+
+    pub(super) fn remove_remote(&self, addr: Addr, gen: slot::Generation) -> Option<T> {
+        let offset = addr.offset() - self.prev_sz;
+        let next = self.push_remote(offset);
+
+        self.slab.get(offset)?.remove(gen, next)
```

Comment:
This logic is highly suspicious to me. You are updating `remote_head` before you are setting the `next` value on the slot. The `insert` path could see the new head w/o the `next` value set.

Reply:
Ah, yeah, you're right — this has been tested in loom but I'll change this logic so that the next value is set first.